march_ethercat/IMC.cpp

```diff
@@ -0,0 +1,137 @@
+#include "ethercat_SDO.h"
+
+// Constructor
+IMC::IMC(std::string name, uint16 number) : Slave(name, number)
+{
+  type = "IMC";
+}
+
+// Publish received data
+void IMC::publish()
+{
+  // Empty for now
+  // Todo: determine what to do with received IMC data
+}
+
+// Map Process Data Object by sending SDOs to the IMC
+bool IMC::PDOmapping()
+{
+  printf("PDO mapping START!\n");
+
+  bool success = true;
+
+  //----------------------------------------
+
+  // clear sm pdos
+  success &= sdo_bit8(number, 0x1C12, 0, 0);
+  success &= sdo_bit8(number, 0x1C13, 0, 0);
+
+  //----------------------------------------
+
+  // clear 1A00 pdo entries
+  success &= sdo_bit32(number, 0x1A00, 0, 0);
+  // download 1A00 pdo entries
+  success &= sdo_bit32(number, 0x1A00, 1, 0x60410010);
+  success &= sdo_bit32(number, 0x1A00, 2, 0x60640020);
+  success &= sdo_bit32(number, 0x1A00, 3, 0x20000010);
+  // download 1A00 pdo count: 3
+  success &= sdo_bit32(number, 0x1A00, 0, 3);
+  //--------------------
+
+  // clear 1A01 pdo entries
+  success &= sdo_bit32(number, 0x1A01, 0, 0);
+  // download 1A01 pdo entries
+  success &= sdo_bit32(number, 0x1A01, 1, 0x20020010);
+  success &= sdo_bit32(number, 0x1A01, 2, 0x20550010);
+  success &= sdo_bit32(number, 0x1A01, 3, 0x20580010);
+  // download 1A01 pdo count: 4
+  success &= sdo_bit32(number, 0x1A01, 0, 3);
+
+  // clear 1A02 pdo entries
+  success &= sdo_bit32(number, 0x1A02, 0, 0);
+  // download 1A02 pdo entries
+  success &= sdo_bit32(number, 0x1A02, 1, 0x60770010);
+  success &= sdo_bit32(number, 0x1A02, 2, 0x207f0010);
+  success &= sdo_bit32(number, 0x1A02, 3, 0x20880020);
+  // download 1A02 pdo count: 1
+  success &= sdo_bit32(number, 0x1A02, 0, 3);
+  // clear 1A03 pdo entries
+  success &= sdo_bit32(number, 0x1A03, 0, 0);
+
+  //----------------------------------------
+
+  // clear 1600 pdo entries
+  success &= sdo_bit32(number, 0x1600, 0, 0);
+  // download 1600 pdo entries
+  success &= sdo_bit32(number, 0x1600, 1, 0x60400010);
+  success &= sdo_bit32(number, 0x1600, 2, 0x607A0020);
+  // download 1600 pdo count: 2
+  success &= sdo_bit32(number, 0x1600, 0, 2);
+
+  //--------------------
+
+  // clear 1601 pdo entries
+  success &= sdo_bit32(number, 0x1601, 0, 0);
+  // clear 1602 pdo entries
+  success &= sdo_bit32(number, 0x1602, 0, 0);
+  // clear 1603 pdo entries
+  success &= sdo_bit32(number, 0x1603, 0, 0);
+
+  //----------------------------------------
+
+  // download 1C12:01 index
+  success &= sdo_bit16(number, 0x1C12, 1, 0x1600);
+  // download 1C12 counter
+  success &= sdo_bit8(number, 0x1C12, 0, 1);
+
+  //--------------------
+
+  // download 1C13:01 index
+  success &= sdo_bit16(number, 0x1C13, 1, 0x1A00);
+  success &= sdo_bit16(number, 0x1C13, 2, 0x1A01);
+  success &= sdo_bit16(number, 0x1c13, 3, 0x1A02);
+  // download 1C13 counter
+  success &= sdo_bit8(number, 0x1C13, 0, 3);
+
+  //----------------------------------------
+
+  printf(success ? "true\n" : "false\n");
```

Comment:
Extremely ambiguous as to where the error happened or what the error is.
Can you think of a way to refactor this so we know which of the 25 `sdo_bit16()` calls throws an error?